doc/decisions/DR024-Versioning-traits-and-specifications-method.md

```diff
@@ -0,0 +1,365 @@
+managementPolicy.
+
+
+### Option 2 - In data
+
+In this conception, version information is sent "over the wire" so to
+speak. Each trait has a version packed into it, which is unpacked on
+the other side, and then used for any version branching.
+
+> **Note**
+>
+> Practically, this would likely be packed into the `traitId` of each
+> trait, something like
+> "`openassetio-mediacreation-v-4:LocatableContent`". This would
+> effectively make each version of a trait a new trait as far as the
+> `openassetio` core library is concerned.
+
+
+### Host side
+```python
+# The host side implementation is almost entirely unchanged from
+# from standard, excepting a new BatchElementError type
+trait_set = {LocatableContentTrait.kId}
+context = self.createTestContext()
+
+result = [None]
+
+def success_cb(idx, traits_data):
+    result[0] = traits_data
+
+def error_cb(idx, batchElementError):
+    if instanceOf(batchElementError, VersionNotSupportedError):
+      incompatible_with_manager_error()
+    ...
+
+self._manager.resolve(
+    [entity_reference], trait_set, ResolveAccess.kRead, context, success_cb, error_cb
+)
+```
+### Manager side
+```python
+# On the manager side, each trait gets unpacked (probably via a utility
+# method provided by traitgen or openassetio), and branching is done
+# on a trait by trait basis
+
+def resolve(
+        self,
+        entityReferences,
+        traitSet,
+        resolveAccess,
+        context,
+        hostSession,
+        successCallback,
+        errorCallback,
+    ):
+
+  # This function is imagined a pre-provided utility, possibly generated
+  # via traitgen, that would error in some way if the trait set does not
+  # contain the locatableContent trait.
+  locatableContent_version = locatableContent_version_from_trait_set(traitSet)
+
+  success_result = TraitsData()
+  if (locatableContent_version == 2):
+    trait = mediacreation.traits.content.LocatableContentTrait(success_result)
+    trait.setUrl(managed_filesystem_locations[ref.toString()])
+    successCallback(idx, success_result)
+  elif (locatableContent_version == 1):
+    trait = mediacreation.legacy.traits.content.LocatableContentTrait_v1(success_result)
+    trait.setLocation(managed_filesystem_locations[ref.toString()])
+    successCallback(idx, success_result)
+  else:
+    raise RuntimeError()
+
+```
+
+#### Pros
+- Both managers and hosts have the option to be backwards compatible.
+- More flexible, traits of different versions can be mixed in the same
+  session, which could allow hosts/managers to perform partial upgrades
+  in the case they are blocked on small specifics.
+- Hosts don't have to interrogate and declare which library versions
+  are in use, which helps especially in the case of second-order
+  dependencies.
+- Seems to fit with existing OpenAssetIO design, especially with
+  `BatchElementError`, as an incompatible version wouldn't necessarily
+  need to error the whole batch.
```

Comment:
[consider] Related point
```suggestion
- Seems to fit with existing OpenAssetIO design, especially with
  `BatchElementError`, as an incompatible version wouldn't necessarily
  need to error the whole batch.
- Mismatched schema versions in use by the host and manager don't 
  entirely preclude them working together from the outset, as long as
  the versions of individual required traits are compatible.
```

Reply:
Isn't this just restating the "more flexible" point at the top of this list?

Reply:
I was trying to contrast this to ahead-of-time negotiation, where it's all-or-nothing. I agree it's more-or-less covered by the first point... I think perhaps the first point is covering too much under the "more flexible" umbrella, and could be split into multiple Pros.

Reply:
Ok, fair enough it is covering quite a lot. I added the point.

Reply:
https://github.com/OpenAssetIO/OpenAssetIO/pull/1267/commits/964b40ca700c948a2ee025aa48ad828b19c67654